File: api/src/main/java/org/algo/invest/controller/MarketDataRestController.java
package org.algo.invest.controller;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.util.*;
import java.util.List;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.algo.invest.core.AppConfig;
import org.algo.invest.dto.ChartDataDto;
import org.algo.invest.dto.QuoteDto;
import org.algo.invest.model.Industry;
import org.algo.invest.model.QuoteRecord;
import org.algo.invest.model.QuoteType;
import org.algo.invest.service.MarketDataService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@RestController
@CrossOrigin({"http://localhost:4200", "http://47.91.78.219:3000"})
public class MarketDataRestController {

    @Autowired
    MarketDataService marketDataService;

    @Autowired
    AppConfig appConfig;

    @RequestMapping(value="/stream/quotes/indices", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<QuoteDto> getIndicesQuotes() {
        Flux<QuoteDto> flux = Mono.just(
                Stream.concat(getQuotes(QuoteType.CURRENCY, Industry.NONE).stream(), getQuotes(QuoteType.INDEX, Industry.NONE).stream()).collect(Collectors.toList()))
                .flatMapMany(Flux::fromIterable);
        return flux.mergeWith(marketDataService.getLatestQuotes().filter(it -> it.getQuoteType() == QuoteType.INDEX || it.getQuoteType() == QuoteType.CURRENCY).map(this::getQuoteDto));
    }

    @RequestMapping(value="/stream/quotes/etf", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<QuoteDto> getEtfQuotes() {
        Flux<QuoteDto> flux = Mono.just(getQuotes(QuoteType.ETF, Industry.NONE)).flatMapMany(Flux::fromIterable);
        return flux.mergeWith(marketDataService.getLatestQuotes().filter(it -> it.getQuoteType() == QuoteType.ETF).map(this::getQuoteDto));
    }

    @RequestMapping(value="/stream/quotes/{industry}", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<QuoteDto> getEcommerceQuotes(@PathVariable String industry) {
        Flux<QuoteDto> flux = Mono.just(getQuotes(QuoteType.EQUITY, Industry.valueOf(industry.toUpperCase()))).flatMapMany(Flux::fromIterable);
        return flux.mergeWith(marketDataService.getLatestQuotes()
                .filter(it -> it.getQuoteType() == QuoteType.EQUITY && appConfig.getSymbolNameMapping().get(it.getSymbol()).getIndustry() == Industry.valueOf(industry.toUpperCase())).map(this::getQuoteDto));
    }

    @RequestMapping(value="/24hOutPerformer", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<QuoteDto> getDailyOutPerformer() {
        Flux<QuoteDto> flux = Mono.just(appConfig.getSymbolNameMapping().keySet().stream().map(s -> marketDataService.getRealtimeStockRecords().get(s))
                .filter(x -> x.getQuoteType() == QuoteType.EQUITY)
                .sorted(Comparator.comparing(QuoteRecord::getRegularMarketChangePercent).reversed())
                .map(this::getQuoteDto).collect(Collectors.toList()).subList(0, 9)).flatMapMany(Flux::fromIterable);
        return flux;
    }

    @RequestMapping(value="/24hUnderPerformer", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<QuoteDto> getDailyUnderPerformer() {
        Flux<QuoteDto> flux = Mono.just(appConfig.getSymbolNameMapping().keySet().stream().map(s -> marketDataService.getRealtimeStockRecords().get(s))
                .filter(x -> x.getQuoteType() == QuoteType.EQUITY)
                .sorted(Comparator.comparing(QuoteRecord::getRegularMarketChangePercent))
                .map(this::getQuoteDto).collect(Collectors.toList()).subList(0, 9)).flatMapMany(Flux::fromIterable);
        return flux;
    }

    @RequestMapping(value="/50dOutPerformer", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<QuoteDto> get50dOutPerformer() {
        Flux<QuoteDto> flux = Mono.just(appConfig.getSymbolNameMapping().keySet().stream().map(s -> marketDataService.getRealtimeStockRecords().get(s))
                .filter(x -> x.getQuoteType() == QuoteType.EQUITY)
                .sorted(Comparator.comparing(QuoteRecord::getFiftyDayAverageChangePercent).reversed())
                .map(this::getQuoteDto).collect(Collectors.toList()).subList(0, 9)).flatMapMany(Flux::fromIterable);
        return flux;
    }

    @RequestMapping(value="/50dUnderPerformer", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<QuoteDto> get50dPerformer() {
        Flux<QuoteDto> flux = Mono.just(appConfig.getSymbolNameMapping().keySet().stream().map(s -> marketDataService.getRealtimeStockRecords().get(s))
                .filter(x -> x.getQuoteType() == QuoteType.EQUITY)
                .sorted(Comparator.comparing(QuoteRecord::getFiftyDayAverageChangePercent))
                .map(this::getQuoteDto).collect(Collectors.toList()).subList(0, 9)).flatMapMany(Flux::fromIterable);
        return flux;
    }

    @RequestMapping(value="/200dOutPerformer", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<QuoteDto> get200dOutPerformer() {
        Flux<QuoteDto> flux = Mono.just(appConfig.getSymbolNameMapping().keySet().stream().map(s -> marketDataService.getRealtimeStockRecords().get(s))
                .filter(x -> x.getQuoteType() == QuoteType.EQUITY)
                .sorted(Comparator.comparing(QuoteRecord::getTwoHundredDayAverageChangePercent).reversed())
                .map(this::getQuoteDto).collect(Collectors.toList()).subList(0, 9)).flatMapMany(Flux::fromIterable);
        return flux;
    }

    @RequestMapping(value="/200dUnderPerformer", produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<QuoteDto> get200dUnderPerformer() {
        Flux<QuoteDto> flux = Mono.just(appConfig.getSymbolNameMapping().keySet().stream().map(s -> marketDataService.getRealtimeStockRecords().get(s))
                .filter(x -> x.getQuoteType() == QuoteType.EQUITY)
                .sorted(Comparator.comparing(QuoteRecord::getTwoHundredDayAverageChangePercent))
                .map(this::getQuoteDto).collect(Collectors.toList()).subList(0, 9)).flatMapMany(Flux::fromIterable);
        return flux;
    }

    @RequestMapping(value="/allQuotes", method=RequestMethod.GET)
    public List<QuoteDto> getAllQuotes() {
        return appConfig.getSymbolNameMapping().keySet().stream().map(s -> marketDataService.getRealtimeStockRecords().get(s))
                .map(this::getQuoteDto).collect(Collectors.toList());
    }

    private List<QuoteDto> getQuotes(QuoteType quoteType, Industry industry){

        return appConfig.getSymbolNameMapping().keySet().stream().map(s -> marketDataService.getRealtimeStockRecords().get(s))
                .filter(x -> x.getQuoteType() == quoteType && appConfig.getSymbolNameMapping().get(x.getSymbol()).getIndustry() == industry)
                .map(this::getQuoteDto).collect(Collectors.toList());
    }

    private QuoteDto getQuoteDto(QuoteRecord quoteRecord) {

        ChartDataDto chartData = new ChartDataDto(
                marketDataService.getHistoryQuotes().containsKey(quoteRecord.getSymbol())
                        && marketDataService.getHistoryQuotes().get(quoteRecord.getSymbol()).size() > 10 ?
                        marketDataService.getHistoryQuotes().get(quoteRecord.getSymbol()).entrySet().stream().collect(lastN(10)).subList(0, 10)
                                .stream().map(x -> x.getValue().getClose().floatValue()).collect(Collectors.toList()) :
                        new ArrayList<Float>());

        if (chartData.getData().size() == 10 && chartData.getData().get(9) != quoteRecord.getRegularMarketPrice().floatValue())
        {
            chartData.getData().remove(0);
            chartData.getData().add(quoteRecord.getRegularMarketPrice().floatValue());
        }

        return new QuoteDto(
                quoteRecord.getSymbol(),
                appConfig.getSymbolNameMapping().get(quoteRecord.getSymbol()).getName(),
                appConfig.getSymbolNameMapping().get(quoteRecord.getSymbol()).getWkn(),
                quoteRecord.getRegularMarketPrice().floatValue(),
                quoteRecord.getRegularMarketChange(),
                quoteRecord.getRegularMarketChangePercent(),
                quoteRecord.getCurrency(),
                quoteRecord.getFullExchangeName(),
                quoteRecord.getMarketCap(),
                quoteRecord.getFiftyDayAverage(),
                quoteRecord.getFiftyDayAverageChangePercent(),
                quoteRecord.getTwoHundredDayAverage(),
                quoteRecord.getTwoHundredDayAverageChangePercent(),
                quoteRecord.getFiftyTwoWeekHigh(),
                quoteRecord.getFiftyTwoWeekHighChangePercent(),
                quoteRecord.getFiftyTwoWeekLow(),
                quoteRecord.getFiftyTwoWeekLowChangePercent(),
                quoteRecord.getQuoteType(),
                Collections.singletonList(chartData));
    }

    @RequestMapping(value="/quote/{symbol}", method=RequestMethod.GET)
    public QuoteDto getQuote(@PathVariable String symbol) {

        QuoteRecord quoteRecord = marketDataService.getRealtimeStockRecords().get(symbol);

        ChartDataDto chartData = new ChartDataDto(
                marketDataService.getHistoryQuotes().containsKey(symbol) &&
                        marketDataService.getHistoryQuotes().get(symbol).size() >= 200 ?
                        marketDataService.getHistoryQuotes().get(symbol).values().stream().collect(lastN(200)).subList(0, 200)
                                .stream().map(x -> x.getClose().floatValue()).collect(Collectors.toList()) :
                        new ArrayList<>());

        if (chartData.getData().size() == 200 && chartData.getData().get(199) != quoteRecord.getRegularMarketPrice().floatValue())
        {
            chartData.getData().remove(0);
            chartData.getData().add(quoteRecord.getRegularMarketPrice().floatValue());
        }

        return new QuoteDto(
                quoteRecord.getSymbol(),
                appConfig.getSymbolNameMapping().get(quoteRecord.getSymbol()).getName(),
                appConfig.getSymbolNameMapping().get(quoteRecord.getSymbol()).getWkn(),
                quoteRecord.getRegularMarketPrice().floatValue(),
                quoteRecord.getRegularMarketChange(),
                quoteRecord.getRegularMarketChangePercent(),
                quoteRecord.getCurrency(),
                quoteRecord.getFullExchangeName(),
                quoteRecord.getMarketCap(),
                quoteRecord.getFiftyDayAverage(),
                quoteRecord.getFiftyDayAverageChangePercent(),
                quoteRecord.getTwoHundredDayAverage(),
                quoteRecord.getTwoHundredDayAverageChangePercent(),
                quoteRecord.getFiftyTwoWeekHigh(),
                quoteRecord.getFiftyTwoWeekHighChangePercent(),
                quoteRecord.getFiftyTwoWeekLow(),
                quoteRecord.getFiftyTwoWeekLowChangePercent(),
                quoteRecord.getQuoteType(),
                Collections.singletonList(chartData));
    }

    public <T> Collector<T, ?, List<T>> lastN(int n) {
        return Collector.<T, Deque<T>, List<T>>of(ArrayDeque::new, (acc, t) -> {
            if(acc.size() == n)
                acc.pollFirst();
            acc.add(t);
        }, (acc1, acc2) -> {
            while(acc2.size() < n && !acc1.isEmpty()) {
                acc2.addFirst(acc1.pollLast());
            }
            return acc2;
        }, ArrayList::new);
    }

    @RequestMapping(value="/log", method=RequestMethod.GET)
    public List<String> getLog() {
        List<String> s_log = new ArrayList<>();
        try{
            FileInputStream fstream = new FileInputStream("logs/bot-app.log");
            BufferedReader br = new BufferedReader(new InputStreamReader(fstream));
            String strLine;
            while ((strLine = br.readLine()) != null)   {
                s_log.add(strLine);
            }
            fstream.close();
        } catch (Exception ignored) {
        }
        return s_log;
    }
}

File: api/src/main/java/org/algo/invest/controller/MarketDataRestController.kt
package org.algo.invest.controller

import org.algo.invest.service.MarketDataService
import org.algo.invest.core.AppConfig
import org.algo.invest.dto.ChartDataDto
import reactor.core.publisher.Flux
import org.algo.invest.dto.QuoteDto
import org.algo.invest.model.Category
import reactor.core.publisher.Mono
import org.algo.invest.model.QuoteType
import org.algo.invest.model.Industry
import java.util.stream.Collectors
import org.algo.invest.model.QuoteRecord
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*
import yahoofinance.histquotes.HistoricalQuote
import java.io.FileInputStream
import java.io.BufferedReader
import java.io.InputStreamReader
import java.lang.Exception
import java.lang.IndexOutOfBoundsException
import java.util.*
import kotlin.streams.asSequence

@RestController
@CrossOrigin("http://localhost:4200", "http://47.91.78.219:3000")
class MarketDataRestController(
    private val marketDataService: MarketDataService,
    private val appConfig: AppConfig)
{
    private val pagesize: Int = 20

    @RequestMapping(value = ["/stream/quotes/{industry}"], produces = [MediaType.TEXT_EVENT_STREAM_VALUE])
    fun quotesByIndustry(@PathVariable industry: String): Flux<QuoteDto> =
        Mono.just(getQuotes(Industry.valueOf(industry.toUpperCase())))
            .flatMapMany { Flux.fromIterable(it) }
            .mergeWith(marketDataService.latestQuotes
                .filter { appConfig.quoteSymbolMetaData[it.symbol]!!.industry == Industry.valueOf(industry.toUpperCase()) }
                .map { getQuoteDto(it) })

    @RequestMapping(value = ["/stream/quotes/{industry}/{category}"], produces = [MediaType.TEXT_EVENT_STREAM_VALUE])
    fun quotesByCategory(@PathVariable industry: String, @PathVariable category: String): Flux<QuoteDto> =
        Mono.just(getQuotes(Industry.valueOf(industry.toUpperCase())))
            .flatMapMany { Flux.fromIterable(it) }
            .mergeWith(marketDataService.latestQuotes
                .filter { it.quoteType == QuoteType.EQUITY
                    && appConfig.quoteSymbolMetaData[it.symbol]!!.industry == Industry.valueOf(industry.toUpperCase())
                    && appConfig.quoteSymbolMetaData[it.symbol]!!.category == Category.valueOf(category.toUpperCase())
                }
                .map { getQuoteDto(it) })

    @RequestMapping(value = ["/stream/quotes/{industry}/{page}"])
    fun page(@PathVariable industry: String, @PathVariable page: Int,
             @RequestParam(required = false) sortProperty: String?, @RequestParam(required = false) sortDirection: String?) =
        getQuotes(Industry.valueOf(industry.toUpperCase()), page, sortProperty, sortDirection)

    @RequestMapping(value = ["/24hOutPerformer"])
    fun dailyOutPerformer() =
        appConfig.quoteSymbolMetaData.keys
            .asSequence()
            .map { marketDataService.realtimeStockRecords.getValue(it) }
            .filter { it.quoteType == QuoteType.EQUITY }
            .sortedByDescending { it.regularMarketChangePercent }
            .map{ getQuoteDto(it) }
            .take(10)
            .toList()

    @RequestMapping(value = ["/24hUnderPerformer"])
    fun dailyUnderPerformer() =
        appConfig.quoteSymbolMetaData.keys
            .asSequence()
            .map { marketDataService.realtimeStockRecords.getValue(it) }
            .filter { it.quoteType == QuoteType.EQUITY }
            .sortedBy { it.regularMarketChangePercent }
            .map{ getQuoteDto(it) }
            .take(10)
            .toList()

    @RequestMapping(value = ["/50dOutPerformer"])
    fun get50dOutPerformer() =
        appConfig.quoteSymbolMetaData.keys
            .asSequence()
            .map { marketDataService.realtimeStockRecords.getValue(it) }
            .filter { it.quoteType == QuoteType.EQUITY }
            .sortedByDescending { it.fiftyDayAverageChangePercent }
            .map{ getQuoteDto(it) }
            .take(10)
            .toList()

    @RequestMapping(value = ["/50dUnderPerformer"])
    fun get50dPerformer() =
        appConfig.quoteSymbolMetaData.keys
                .asSequence()
                .map { marketDataService.realtimeStockRecords.getValue(it) }
                .filter { it.quoteType == QuoteType.EQUITY }
                .sortedBy { it.fiftyDayAverageChangePercent }
                .map{ getQuoteDto(it) }
                .take(10)
                .toList()

    @RequestMapping(value = ["/200dOutPerformer"])
    fun get200dOutPerformer() =
        appConfig.quoteSymbolMetaData.keys
            .asSequence()
            .map { marketDataService.realtimeStockRecords.getValue(it) }
            .filter { it.quoteType === QuoteType.EQUITY }
            .sortedByDescending { it.twoHundredDayAverageChangePercent }
            .map{ getQuoteDto(it) }
            .take(10)
            .toList()

    @RequestMapping(value = ["/200dUnderPerformer"])
    fun get200dUnderPerformer() =
        appConfig.quoteSymbolMetaData.keys.stream()
            .asSequence()
            .map { marketDataService.realtimeStockRecords.getValue(it) }
            .filter { it.quoteType === QuoteType.EQUITY }
            .sortedBy { it.twoHundredDayAverageChangePercent }
            .map{ getQuoteDto(it) }
            .take(10)
            .toList()

    private fun getQuotes(industry: Industry): List<QuoteDto> =
        appConfig.quoteSymbolMetaData.keys
            .asSequence()
            .map { marketDataService.realtimeStockRecords.getValue(it) }
            .filter{ appConfig.quoteSymbolMetaData.getValue(it.symbol!!).industry === industry }
            .map{ getQuoteDto(it) }
            .toList()

    private fun getQuotes(industry: Industry, page: Int, sortProperty: String?, sortDirection: String?): List<QuoteDto> =
        try {
            appConfig.quoteSymbolMetaData.keys
                .asSequence()
                .map { marketDataService.realtimeStockRecords.getValue(it) }
                .filter{ appConfig.quoteSymbolMetaData.getValue(it.symbol!!).industry === industry }
                .sortedWith { a, b -> compare(a,b, sortProperty, sortDirection) }
                .chunked(pagesize)
                .elementAt(page)
                .map{ getQuoteDto(it) }
                .toList()
        } catch (exception: IndexOutOfBoundsException){
            emptyList()
        }

    private fun getQuoteDto(quoteRecord: QuoteRecord): QuoteDto {
        val chartData = ChartDataDto(
            if (marketDataService.historyQuotes.containsKey(quoteRecord.symbol) && marketDataService.historyQuotes.getValue(quoteRecord.symbol).size > 10)
                marketDataService.historyQuotes[quoteRecord.symbol]!!.entries.toList().takeLast(10)
                    .stream().map { x: Map.Entry<Calendar?, HistoricalQuote> -> x.value.close.toFloat() }
                    .collect(Collectors.toList())
            else ArrayList())
        if (chartData.data?.size == 10 && chartData.data[9] != quoteRecord.regularMarketPrice.toFloat()) {
            Collections.rotate(chartData.data, -1)
            chartData.data.removeLast()
            chartData.data.add(quoteRecord.regularMarketPrice.toFloat())
        }
        return quoteDto(quoteRecord, chartData)
    }

    @RequestMapping(value = ["/quote/{symbol}"], method = [RequestMethod.GET])
    fun getQuote(@PathVariable symbol: String): QuoteDto {
        val quoteRecord = marketDataService.realtimeStockRecords.getValue(symbol)
        val chartData = ChartDataDto(
            if (marketDataService.historyQuotes.containsKey(symbol) && marketDataService.historyQuotes.getValue(symbol).size >= 200)
                marketDataService.historyQuotes.getValue(symbol).values.toList().takeLast(200)
                    .stream().map { it.close.toFloat() }
                    .collect(Collectors.toList())
            else ArrayList()
        )
        if (chartData.data?.size == 200 && chartData.data[199] != quoteRecord.regularMarketPrice.toFloat()) {
            Collections.rotate(chartData.data, -1)
            chartData.data.removeLast()
            chartData.data.add(quoteRecord.regularMarketPrice.toFloat())
        }
        return quoteDto(quoteRecord!!, chartData)
    }

    @get:RequestMapping(value = ["/log"], method = [RequestMethod.GET])
    val log: List<String>
        get() {
            val sLog: MutableList<String> = ArrayList()
            try {
                val stream = FileInputStream("logs/bot-app.log")
                val br = BufferedReader(InputStreamReader(stream))
                var strLine: String
                while (br.readLine().also { strLine = it } != null) {
                    sLog.add(strLine)
                }
                stream.close()
            } catch (ignored: Exception) {
            }
            return sLog
        }

    private fun quoteDto(quoteRecord: QuoteRecord, chartData: ChartDataDto) =
        QuoteDto(
            appConfig.quoteSymbolMetaData.getValue(quoteRecord.symbol!!).category,
            quoteRecord.symbol,
            appConfig.quoteSymbolMetaData.getValue(quoteRecord.symbol).name,
            appConfig.quoteSymbolMetaData.getValue(quoteRecord.symbol).wkn,
            quoteRecord.regularMarketPrice.toFloat(),
            quoteRecord.regularMarketChange,
            quoteRecord.regularMarketChangePercent,
            quoteRecord.currency,
            quoteRecord.fullExchangeName,
            quoteRecord.marketCap,
            quoteRecord.fiftyDayAverage,
            quoteRecord.fiftyDayAverageChangePercent,
            quoteRecord.twoHundredDayAverage,
            quoteRecord.twoHundredDayAverageChangePercent,
            quoteRecord.fiftyTwoWeekHigh,
            quoteRecord.fiftyTwoWeekHighChangePercent,
            quoteRecord.fiftyTwoWeekLow,
            quoteRecord.fiftyTwoWeekLowChangePercent,
            quoteRecord.quoteType,
            listOf(chartData)
        )

    private fun compare(a: QuoteRecord, b: QuoteRecord, sortProperty: String?, sortDirection: String?) =
        when (sortProperty){
            "name" ->
                when {
                    a.symbol.toString() > b.symbol.toString() -> if (sortDirection == "desc") -1 else 1
                    a.symbol.toString() < b.symbol.toString() -> if (sortDirection == "desc") 1 else -1
                    else -> 0
                }
            "dayChangePercent" ->
                when {
                    a.regularMarketChangePercent > b.regularMarketChangePercent -> if (sortDirection == "desc") -1 else 1
                    a.regularMarketChangePercent < b.regularMarketChangePercent -> if (sortDirection == "desc") 1 else -1
                    else -> 0
                }
            "fiftyDayAverageChangePercent" ->
                when {
                    a.fiftyDayAverageChangePercent > b.fiftyDayAverageChangePercent -> if (sortDirection == "desc") -1 else 1
                    a.fiftyDayAverageChangePercent < b.fiftyDayAverageChangePercent -> if (sortDirection == "desc") 1 else -1
                    else -> 0
                }
            "twoHundredDayAverageChangePercent" ->
                when {
                    a.twoHundredDayAverageChangePercent > b.twoHundredDayAverageChangePercent -> if (sortDirection == "desc") -1 else 1
                    a.twoHundredDayAverageChangePercent < b.twoHundredDayAverageChangePercent -> if (sortDirection == "desc") 1 else -1
                    else -> 0
                }
            "fiftyTwoWeekLowChangePercent" ->
                when {
                    a.fiftyTwoWeekLowChangePercent > b.fiftyTwoWeekLowChangePercent -> if (sortDirection == "desc") -1 else 1
                    a.fiftyTwoWeekLowChangePercent < b.fiftyTwoWeekLowChangePercent -> if (sortDirection == "desc") 1 else -1
                    else -> 0
                }
            "fiftyTwoWeekHighChangePercent" ->
                when {
                    a.fiftyTwoWeekHighChangePercent > b.fiftyTwoWeekHighChangePercent -> if (sortDirection == "desc") -1 else 1
                    a.fiftyTwoWeekHighChangePercent < b.fiftyTwoWeekHighChangePercent -> if (sortDirection == "desc") 1 else -1
                    else -> 0
                }
            "marketCap" ->
                when {
                    a.marketCap > b.marketCap -> if (sortDirection == "desc") -1 else 1
                    a.marketCap < b.marketCap -> if (sortDirection == "desc") 1 else -1
                    else -> 0
                }
            else -> 0
        }
}